model/state_model_recurse.jl

```diff
@@ -0,0 +1,760 @@
+
+        d1 = Dict(k => filter_with(df, (yr = year,); extrapolate = false, drop = true) for (k,df) in d)
+        d1 = Dict(k => convert_type(Dict, fill_zero(set, df)) for (k,df) in d1)
+        return (d1, set, idx)
+end
+
+function _model_input(year::UnitRange{Int64}, d::Dict{Symbol,DataFrame}, set::Dict, idx::Dict = Dict())
+        return _model_input(ensurearray(year), d, set, idx)
+end
+
+"""
+    _model_set!(d::Dict{Symbol,DataFrame}, set::Dict, idx::Dict)
+This function returns subsets intended to limit the size of the model by including only
+non-zero values when mapping zero-profit and market-clearing conditions.
+
+# Arguments
+- `d::Dict{Symbol,DataFrame}` of DataFrames containing the model data.
+- `set::Dict` of Arrays describing region, sector, final demand, etc.
+- `idx::Dict` of parameter indices
+
+# Returns
+- `d::Dict{Symbol,Dict}` of model data with all zeros filled
+- `set::Dict` of model indices, with added arrays of tuples indicating which (region,sector)
+    or (region,good) combinations are relevant to the zero-profit and market-clearing
+    conditions.
+- `idx::Dict` of indices, updated to include those used to define the newly-added sets.
+"""
+function _model_set!(d::Dict{Symbol,DataFrame}, set::Dict, idx::Dict)
+    (set[:A], idx[:A]) = nonzero_subset(d[:a0] + d[:rx0])
+    (set[:Y], idx[:Y]) = nonzero_subset(combine_over(d[:ys0], :g))
+    (set[:X], idx[:X]) = nonzero_subset(d[:s0])
+    (set[:PA], idx[:PA]) = nonzero_subset(d[:a0])
+    (set[:PD], idx[:PD]) = nonzero_subset(d[:xd0])
+    (set[:PK], idx[:PK]) = nonzero_subset(d[:kd0])
+    (set[:PY], idx[:PY]) = nonzero_subset(d[:s0])
+    return (set, idx)
+end
+
+
+
+############
+# LOAD DATA
+############
+
+#SLiDE data needs to be built or point to pre-existing build directory
+#can pass a name (d, set) = build_data("name_of_build_directory")
+!(@isdefined(d_in) && @isdefined(set_in)) && ((d_in, set_in) = build_data("state_model"))
+d = copy(d_in)
+set = copy(set_in)
+
+#Specify benchmark year and end year for dynamic model time horizon
+bmkyr = 2016
+#endyr = 2018
+
+#Define range of years in time horizon
+#years = bmkyr:endyr
+
+#Load slide data and time horizon to produce model data and appropriate time-indexed subsets
+(sld, set, idx) = _model_input(bmkyr, d, set)
+
+#last year is the maximum of all years
+# years_last = maximum(years)
+
+#Create boolean indicators of first and last year of time horizon
+# bool_firstyear = Dict()
+# bool_lastyear = Dict()
+# for t in years
+#         if t!=years_last
+#                 push!(bool_lastyear,t=>0)
+#         else
+#                 push!(bool_lastyear,t=>1)
+#         end
+
+#         if t!=bmkyr
+#                 push!(bool_firstyear,t=>0)
+#         else
+#                 push!(bool_firstyear,t=>1)
+#         end
+# end
+
+###############
+# -- SETS --
+###############
+
+regions = set[:r]
+sectors = set[:s]
+goods = set[:g]
+margins = set[:m]
+goods_margins = set[:gm]
```

Comment:
```
julia> @time @NLparameter(cge, ys0alt[r in regions, s in sectors, g in goods] == sld[:ys0][r,s,g]);
  0.509867 seconds (2.54 M allocations: 77.727 MiB)

julia> @time @NLparameter(cge, ys0[r in set[:r], s in set[:s], g in set[:g]] == sld[:ys0][r,s,g]);
  0.308344 seconds (2.54 M allocations: 76.724 MiB)
```

It appears that using `set[:r]` is faster than using `regions = set[:r]`.  Unsure why? Maybe because `regions` adds an additional reference.

`set[:r]` is consistent with the rest of SLiDE. Though there are times where it is redefined within functions, but I think that is only when they are dataframe column names.

Reply:
```
@time set[:r]
@time regions
```

With set[:r] you have to look something up. With regions, the looking up has already been done.